### ISSUE TITLE
some questions about storage

### ISSUE THREAD

**Issue by knowmara** [vespa-engine/vespa#4704]
hi,i have some questions about storage of vespa,listed as follow:
1  When a large amount of data is imported, buckets will always be in the process of splitting. According to the bucket distribution algorithm, a large number of buckets will be frequently migrated in the cluster, consuming a large amount of IO and bandwidth,which may cause performance problems?
2 The index of a bucket (assumed to be seachable copy) will eventually have only one inverted index unlike lucene will have index into multiple segments,  I understand right?
3 When writing data, the distributor writes data to all copys of the bucket selected based on the distribution algorithm at the same time, or writes to the bucket copy (which may be called primary) first and then writes to the remaining copys by the primary copy?
4 The document of Proton maintenance jobs says disk index fusion only requires very little memory consumption, but,in my opinion index fusion at least need to read the index into memory, thus it requires a lot of memory.I understand Right?
5 Storage node choice is specified in the content part of service.xml, when a new node to join, only modify the service.xml on the storage node configuration will allow the bucket to use the newly added node, I understand Right?
6 why only active index in ready are searchable,all searchable copies are consistent, is not it better to provide a search service for all copies?

thanks

**Comment by vekterli** [comment]
Hi knowmara, I’ll try to answer some of your questions:
1. When using random document distribution (i.e. document IDs do not have a `n=` or `g=` component), splitting will normally _not_ require moving buckets across nodes. This is due to the distribution algorithm having an affinity for keeping a bucket on the same node until it has been split beyond a certain level. The rationale is that since documents have already been (pseudo)randomly distributed across nodes, redistributing them again will not improve the overall distribution. Streaming search is a somewhat different story (by design), I can go into more details here if you want.
2. Correct in the sense that all buckets on a node share the same indexes. A bucket contains a set of documents (possibly of different types) and these have fields that may be covered by indexes. There is no connection between the number of buckets and the number of indexes, which is why adding more buckets will not cause a slowdown by e.g. requiring additional index fsyncs.
3. Client write operations are fanned out by the distributor in parallel to all replicas, they are not chained from the primary.
4. @geirst or @baldersheim can you answer this?
5. Correct, the system will automatically and immediately begin redistribution and (re-)indexing of bucket replicas once a node has been added to (or removed from) services.xml.
6. A bucket replica being `active` means it partakes in searches and may be part of search results. A bucket that is _not_ `active` but _is_ `ready` (indexed) due to `searchable-copies` > 1 does _not_ partake in searches. Since (generally) all indexed replicas of a bucket are consistent, having more than one replica considered active would lead to duplicates in the search result. This is because searches are executed in parallel across all nodes, and having the same bucket be part of searches on more than one node means it would also be part of results from both nodes. The inactive, ready replica stands ready to be activated within seconds if the node holding the previously active replica goes down.

Please let me know if this did not answer your questions fully, or if you want more detailed information on any of the points!

**Comment by baldersheim** [comment]
Index fusion does not need the whole index in memory. It acts as a simple N way merge. So we just stream the indexes creating a new fusioned one.

**Comment by knowmara** [comment]
Thanks for your answer, @baldersheim 
I think so too,but i have some doubt.Assumed that there are 5 documents(doc id is 1,2,3,4,5), the disk inverted index is:

> "a" -->1,3
"b"-->4
"c"-->2,5
and flushed memory inverted index is:

> "a"-->2,5
"b"-->3
"c"-->1,4
In my opinion,inverted index is in dictionary order,first read **"a" -->1,3** of disk index and **"a"-->2,5** of flushed memory index,then merge into **"a"-->1,2,3,5**,at last write it into disk,and so on.In this way,we can not delete the old index **"a"-->1,3**,because we always can not see the full view of index in memory which included the deleting semantic.
Do i understand right?

**Comment by baldersheim** [comment]
Ah, now I see what you mean.
Fusion is just the process of merging 2 or more indexes to one a new single index. This is the process running in the background. This has very low memory impact as it is a purely sequential merge.

However switching in the new index for serving is different which takes place after fusion is done is different.
Given the that we start with index A,B,C on disk and the current one in memory M1.
A, B, C are frozen. M1 are receiving updates.
When we decide to perform fusion we will start a background job that merges A, B, C and produce D. When fusion completes we are still serving from the set {A,B,C,M1}.
The we will initiate the process of switching serving from {A, B, C, M1} to {D, M1}.
The switch itself is atomic. Dropping of {A, B, C} takes place when all queries associated with it are completed which is normally within a second. The loading of D1 from disk to memory is lazy and on demand. This means that the first queries towards {D, M1} will incur extra latency due to pulling in from disk to memory. However you can counter that by specifying that you want it to be warmed up. This means that before doing the atomic swap we will in a specified period fx 30 seconds, send the queries to {D1} so it will be warmed up. And then the latency cost of the switch will be minor.

Keep in mind that and index consists of many independent files. For each field there is are different sets of files. One for the dictionary, and multiple for the posting/payload. Normally the dictionary is hot and fully mapped in while only parts of the posting files will end up in memory. There are indexed a lot of terms that are never requested during query time. Typically 15-30% of an index are mapped in after a minute or 2.

**Comment by vekterli** [comment]
1. Here’s the basic outline of how splits are performed:

   1. The distributor observes that a bucket has grown beyond the configurable thresholds (too many documents and/or too much space used)
   2. A bucket split command is sent in parallel to all replicas
   3. All replica nodes perform the split on their locally stored bucket, resulting in (generally) two new buckets that are at a deeper split level than the parent bucket they were split from. The common case is that each child bucket receives about 50% of the documents of the parent bucket. Note that _nothing is moved across nodes as part of the splitting operation itself_.
   4. The distributor updates its bucket database with the newly split buckets and removes the source bucket from its internal state
   5. At this point, it’s possible that replicas of one or both of the new buckets may be in a location that the [pseudo-random distribution algorithm](http://docs.vespa.ai/documentation/elastic-vespa.html#bucket-distribution-algorithm) considers non-ideal. If this is the case, the replicas will be moved to their new ideal locations as a background operation (a “merge” operation).
As for streaming search, an increase in the number of buckets will indeed lead to a greater level of parallelism. Since data is streamed from disk and a search potentially requires reading a lot of data, we want to maximize the number of nodes that can do this in parallel. Buckets with documents used for streaming search are therefore split in such a way that they will be moved much more aggressively to additional nodes for this very reason.
2. Are you referring to the computed size of a bucket that is used for deciding if it should be split or not? The value used is the sum of the serialized binary size of the documents contained in the bucket, and do not factor in overhead caused by index structures.
3. _Indexed_ search requests are actually not routed through the distributor at all (partly for historical reasons, partly for performance due to wanting as few indirections as possible). This means that when a search is executed against the search nodes, it happens on a code path with no knowledge of what replicas exist for a bucket. This means that explicit activation is required to avoid duplicates.
_Streaming_ search requests _are_ routed through the distributor to specific bucket replicas in parallel, which is closer to what you're describing. Here, activation is not needed to avoid duplicates.

**Comment by knowmara** [comment]
Benefit a lot, thanks for your guidance, @vekterli @baldersheim,thanks~~
